openshift inventory plugin: fix exception when auth fails

pilou- opened the pull request:
##### SUMMARY

Fix `ForbiddenError` object has no attribute `message`:

    [WARNING]:  * Failed to parse test.yml with openshift plugin: 'ForbiddenError' object has no attribute 'message'
     File "ansible/lib/ansible/inventory/manager.py", line 270, in parse_source
       plugin.parse(self._inventory, self._loader, source, cache=cache)
     File "ansible/lib/ansible/plugins/inventory/openshift.py", line 122, in parse
       self.setup(config_data, cache, cache_key)
     File "ansible/lib/ansible/module_utils/k8s/inventory.py", line 58, in setup
       self.fetch_objects(connections)
     File "ansible/lib/ansible/module_utils/k8s/inventory.py", line 250, in fetch_objects
       super(OpenShiftInventoryHelper, self).fetch_objects(connections)
     File "ansible/lib/ansible/module_utils/k8s/inventory.py", line 81, in fetch_objects
       namespaces = self.get_available_namespaces(client)
     File "ansible/lib/ansible/module_utils/k8s/inventory.py", line 95, in get_available_namespaces
       raise K8sInventoryException('Error fetching Namespace list: {0}'.format(exc.message))

Don't try to get `message` attribute from:
- `K8sInventoryException` instances
- `Exception` instances
- `KubernetesException` instances (because `KubernetesException` can be `Exception`)

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
lib/ansible/plugins/inventory/openshift.py

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes -->
```
ansible 2.8.0.dev0 (devel af40d8c2a5) last updated 2018/09/19 01:46:21 (GMT +200)
```

pilou- committed:
openshift inventory: fix exception when auth fails

Fix 'ForbiddenError' object has no attribute 'message':

    [WARNING]:  * Failed to parse test.yml with openshift plugin: 'ForbiddenError' object has no attribute 'message'
     File "ansible/lib/ansible/inventory/manager.py", line 270, in parse_source
       plugin.parse(self._inventory, self._loader, source, cache=cache)
     File "ansible/lib/ansible/plugins/inventory/openshift.py", line 122, in parse
       self.setup(config_data, cache, cache_key)
     File "ansible/lib/ansible/module_utils/k8s/inventory.py", line 58, in setup
       self.fetch_objects(connections)
     File "ansible/lib/ansible/module_utils/k8s/inventory.py", line 250, in fetch_objects
       super(OpenShiftInventoryHelper, self).fetch_objects(connections)
     File "ansible/lib/ansible/module_utils/k8s/inventory.py", line 81, in fetch_objects
       namespaces = self.get_available_namespaces(client)
     File "ansible/lib/ansible/module_utils/k8s/inventory.py", line 95, in get_available_namespaces
       raise K8sInventoryException('Error fetching Namespace list: {0}'.format(exc.message))

Don't try to get 'message' attribute from:
- K8sInventoryException instances
- Exception instances
- KubernetesException instances (because KubernetesException can be
  Exception)

willthames commented:
I'd rather not use `exc` for `DynamicApiException`, it's very verbose - can we use `json.loads(exc.body)['message']` instead?

(would be good to have a helper function to do that, I want to make the same change in k8s/raw.py too)

willthames commented:
@pilou- that looks great. I'm not clear why we lose the backtrace though, and if there's anything we can do to get it back (is it just that it was part of e.message?)

We can presumably use the `traceback` method and maybe put that in as a keyword arg to the exception (a bit like how `fail_json_aws` works - perhaps a `fail_json_k8s` would look pretty similar?)

pilou- committed:
move k8s/OpenShift inventory plugin dedicated code

inventory plugin specific code should not be located in
lib/ansible/module_utils directory. Then ansible.utils methods can be
reused (for example Display).

pilou- commented:
About the traceback: it is not possible to use something like `fail_json_aws` because there is no `AnsibleModule` instance there. In fact `module.utils.Display` should be used in order to be more verbose when debug is enabled. That's why I moved all code specific to inventory only from `ansible.module_utils.k8s` to `ansible.plugins.inventory.k8s`.

Now `Display` can be used and traceback is displayed when `-vvv` switch is used.
Note there isn't any license problem since `ansible.module_utils.k8s` is GPL.

fabianvf commented on lib/ansible/plugins/inventory/openshift.py:
```diff
@@ -108,15 +108,86 @@
-        cache_key = self._get_cache_prefix(path)
-        config_data = self._read_config_data(path)
-        self.setup(config_data, cache, cache_key)
+    helper = None
```
what is this for?

pilou- replied:
It comes from: https://github.com/ansible/ansible/blob/db5b65f5d618859ebb23c815797f7d88057c7fea/lib/ansible/module_utils/k8s/inventory.py#L246 Not used since 4d77878654e867c23d5f6c61422bdae7120393bc: I removed it.

pilou- committed:
Remove unused class variables 'helper'

unused since 4d77878654e867c23d5f6c61422bdae7120393bc.

fabianvf commented:
:shipit:

willthames commented on lib/ansible/plugins/inventory/k8s.py:
```diff
@@ -108,15 +108,245 @@
+class InventoryModule(BaseInventoryPlugin, Constructable, Cacheable, K8sAnsibleMixin):
     NAME = 'k8s'
 
+    transport = 'kubectl'
```
What is this for? kubectl is not used at all by ansible, afaik

pilou- replied:
I didnt' add it, it comes from: https://github.com/ansible/ansible/blob/f73996e6ebc25a453caef3631558fca6e1af4a4e/lib/ansible/module_utils/k8s/inventory.py#L40

It is used there in order to define connection plugin (`kubectl` or `oc`) which will be used with the generated inventory:
https://github.com/ansible/ansible/blob/f73996e6ebc25a453caef3631558fca6e1af4a4e/lib/ansible/module_utils/k8s/inventory.py#L161-L167

maxamillion replied:
`kubectl` is a valid connection plugin, without looking into the code further I assume that's being used as the transport backend in that inventory plugin.

https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/connection/kubectl.py

maxamillion replied:
https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/connection/kubectl.py

pilou- replied:
Yeah, `ansible_connection` [selects](https://docs.ansible.com/ansible/latest/user_guide/intro_inventory.html#non-ssh-connection-types) the connection plugin :)

fabianvf commented:
bot_status

fabianvf commented:
shipit

maxamillion commented:
rebuild_merge
